### ISSUE TITLE
Haproxy doesn't recover NOLB status after a server went down

### ISSUE THREAD

**Issue by yoshz** [haproxy/haproxy#934]
## Detailed description of the problem

We have various servers configured with **http-check disable-on-404** which get a NOLB status when the check endpoint responds with a 404.
However when the server is temporary down and comes back up again, Haproxy doesn't change the status from DOWN to NOLB.

## Expected behavior
After a server was DOWN, the server should go back to NOLB again when the check endpoint responds with a 404.

## Steps to reproduce the behavior

1. Configure a backend with disable-on-404 enabled. 1 server is master and will return 200, the other will return 404.
```
listen redis
  bind :6379 
  mode tcp
  default-server inter 5s
  http-check disable-on-404
  http-check expect status 200
  option httpchk GET /
  server redis-01 10.0.0.201:6379 check port 8080
  server redis-02 10.0.0.202:6379 check port 8080
  server redis-03 10.0.0.202:6379 check port 8080
```
2. Start haproxy and check the dashboard that 1 server has status UP and 2 servers have status NOLB.
3. Shutdown one of the servers that have status NOLB.
4. Check the dashboard and see that the status of that server is changed to DOWN.
5. Start the server again and wait until it is running.
6. Check the dashboard and see that the status of that server is still DOWN but LastChk has the value of "L7STS/404 in 7ms"
7. Reload haproxy
8. Check the dashboard and see that the status of that server is back to NOLB.

## Do you have any idea what may have caused this?
Seems like a bug that causes not to recover to NOLB status after a server went down.

## Do you have an idea how to solve the issue?

No.

## What is your configuration?

See above.

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 1.8.24 2020/02/15
Copyright 2000-2020 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace
```

```
Linux haproxy 3.10.0-1127.10.1.el7.x86_64 #1 SMP Wed Jun 3 14:28:03 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

## Additional information (if helpful)

Also had this issue with version 1.5 and 2.0 (don't know the patch version anymore).

**Comment by yoshz** [comment]
Does anybody have the same problem?

**Comment by capflam** [comment]
I'm not sure, but it seems to be intentional. A server is set to NOLB state if a 404 is received while it is in UP state. But once in DOWN state, it must be UP again to make the `disable-on-404` option evaluated again.

**Comment by yoshz** [comment]
@capflam After DOWN state the server is actually UP again, the healthcheck endpoint replies with a 200. Haproxy just doesn't trigger a state change from DOWN to NOLB again.

**Comment by capflam** [comment]
I don't understand. If the server replies with 200, its server state should be UP. It is then set to NOLB from UP if it replies a 404. But it must be UP first. From DOWN state, it can't be set to NOLB. So you must have DOWN -> UP (200) -> NOLB (404). If it works from a reload, it is because servers are UP on startup by default.

It is how it works for now. And as said, it seems to be intentional. I missed something ?

**Comment by yoshz** [comment]
Let me explain the use case above for the redis servers.
We have 3 redis servers, 1 master and 2 slaves.
The healthcheck script on each server replies a 404 when it is a slave and replies 200 when it is a master.
Haproxy is configured with the disable-on-404 so all slaves get the NOLB status instead of DOWN, which works great with the monitoring tools as they only report about DOWN status.
When the master goes down, a slave will take will be promoted to master and starts replying 200.

This all works great until we put one slave into maintenance. Haproxy changes the slave status to DOWN.
When we finished maintenance and the slave is replying 404's again.It will never go to UP status as it will never reply with a 200.
Haproxy will keep reporting the server as DOWN instead of NOLB.

In my opinion Haproxy must handle 404's to change the status back to NOLB after a node was DOWN.

**Comment by capflam** [comment]
So, I fully understand what you said :) And It is the expected behaviour of disable-on-404. It just switches an UP server to NOLB state instead of DOWN if it replies a 404. But it does not move a DOWN server to NOLB state. It was designed this way (see  48494c0c5c). Honestly, I don't really know why. I'll ask @wtarreau. It is a pretty old feature.

**Comment by capflam** [comment]
So it is confirmed. It works as expected. Idea of the `disable-on-404` option is to drain traffic on a server before shutting it down, for maintenance for instance. When the server is restarted, if it still replies a 404, it may be a problem and not expected. So it remains in DOWN state to be sure to not forward traffic on a not fully available server.

I understand that an HTTP check is probably easier to setup. However what you performed is closer to an agent-check than a health-check.  It may be a good feature to extend the agent-check to support the HTTP.

Another solution is to use a tcp-check ruleset instead of an http-check one, requesting redis itself to get its replication info. Something like (from the doc):
```
         option tcp-check
         tcp-check comment PING\ phase
         tcp-check send PING\r\n
         tcp-check expect string +PONG
         tcp-check comment role\ check
         tcp-check send info\ replication\r\n
         tcp-check expect string role:master
         tcp-check comment QUIT\ phase
         tcp-check send QUIT\r\n
         tcp-check expect string +OK
```
But the slaves will be marked as DOWN.

**Comment by yoshz** [comment]
Thanks for your reply and looking into it.

Changing the HTTP check to a TCP check still results in DOWN status and unnecessary monitoring alerts.
This way we have to change the monitoring solution to ignore DOWN status for specific cases which seems wrong.

Isn't there a different way in Haproxy to only activate one server in a pool of servers based on the result of check script?
Or is it possible to add an extra configuration option for our use case?

**Comment by capflam** [comment]
You may match on the check status in addition on the server status. In my example, slave servers are set to DOWN with a check status set to L7RSP. But a connection failure will be reported with a check status set to L4CONN.

Another solution is to use a simple redis health-check to known if your server is UP or DOWN and use an agent check to set slaves to DRAIN mode and master to READY mode. The agent is not an HTTP service but it does exactly what you want.

**Comment by wtarreau** [comment]
Christopher, now that http-checks are implemented using tcp-checks at the lower level, are the tcp-checks the ones setting the server to drain mode or is there still some code specific to the http disable-on-404 ? I'm asking because if using tcp-check internal rules it's technically possible to make a server enter or leave the drain administrative state, maybe it could be useful to figure a way to expose these one way or another. I know it could be particularly painful and full of corner cases, I'm just asking about your feeling on this.

**Comment by capflam** [comment]
All checks (agent and health) are based on tcp-checks. So yes, drain/maint mode are set from tcp-checks, and disable-on-404 is also handled from tcp-checks. One possible way is to add parameters on tcp-check/http-check expect line. But the harder is probably to keep a consistent state on transition.

**Comment by capflam** [comment]
I think adding http support on agent-check or refine its format in a way or another may be a good idea. But using health-checks to do administrative task is confusing.

**Comment by wtarreau** [comment]
I totally agree in general, but given that it was already implemented for HTTP I also see valid use cases (which would definitely be better addressed by an HTTP agent).

**Comment by capflam** [comment]
disable-on-404 is used to set the server in NOLB state. It is possible to add enable-on-404 option. But let me test something. Maybe I'm smarter than expected ...

**Comment by wtarreau** [comment]
You're right, we could argue that NOLB is not an administrative state and that technically it could be seen as a death warning. (you could want to return it only based on the load for example).

**Comment by capflam** [comment]
Ok, there is a way to specify the check status, on success, on error and on timeout. For now, unfortunately 7OKC (used to switch the serveur is NOLB state) can only be used on success. But with a very small patch, it could be possible to do such things:
```
    option httpchk GET /isalive
    http-check expect status 200,404
    http-check expect status 200 error-status L7OKC
```
And it is probably doable to add a special error/tout status to continue on the next rule to write:
```
    option httpchk GET /isalive
    http-check expect status 200 error-status CONT
    http-check expect status 404 ok-status L7OKC
```

And of course more scripting rules may be added. 

So I'm not so smart. But almost.

**Comment by wtarreau** [comment]
Indeed, you're smarter than you seem, from now on I will have greater respect for you!

**Comment by yoshz** [comment]
@capflam Sounds great. Any plans on implementing this? :)

**Comment by capflam** [comment]
On second thoughts, only the first version is possible. Or I should also add a way to interrupt the expect rules evaluation on success. But such scripting options must be evaluated first because there are probably many other ways to do so and maybe more convenient.

So, about the L7OKC status on error, I guess I can easily add it on my spare time. It is not an huge work. However, note it may only be backported as far as 2.2, but it must be evaluated first. So, firstly, it will only be available on the 2.4-dev.

**Comment by capflam** [comment]
FYI, it was pushed in 2.4-dev.
